xtask/src/main.rs

```diff
@@ -0,0 +1,282 @@
+    fs::{self, File},
+    path::Path,
+};
+
+fn print_help() {
+    eprintln!("cargo xtask [subcommand]");
+    eprintln!();
+    eprintln!("Supported subcommands are:");
+    eprintln!("\tbindgen\tGenerate rust-bindgen for grpcio-sys package");
+    eprintln!("\tsubmodule\tInit necessary submodules for compilation");
+    eprintln!("\tclang-lint\tLint cpp code in grpcio-sys package");
+    eprintln!("\tcodegen\tGenerate rust code for all protocols");
+}
+
+fn cargo() -> Command {
+    match env::var("CARGO") {
+        Ok(s) => Command::new(s),
+        Err(_) => {
+            eprintln!("no CARGO in environment variables, please invoke the binary by cargo xtask");
+            process::exit(1);
+        }
+    }
+}
+
+fn exec(c: &mut Command) {
+    if let Err(e) = c.status() {
+        eprintln!("failed to execute {:?}: {}", c, e);
+        process::exit(-1);
+    }
+}
+
+fn find_default_arch() -> String {
+    let s = String::from_utf8(
+        Command::new("rustc")
+            .args(&["--print", "cfg"])
+            .output()
+            .unwrap()
+            .stdout,
+    )
+    .unwrap();
+    for l in s.lines() {
+        if let Some(arch) = l.strip_prefix("target_arch=") {
+            if !arch.is_empty() {
+                return arch[1..arch.len() - 1].to_string();
+            }
+        }
+    }
+    panic!("arch not found in {:?}", s);
+}
+
+fn remove_match(data: &str, pattern: impl Fn(&str) -> bool) -> String {
+    let mut target = String::with_capacity(data.len());
+    for l in data.lines() {
+        if pattern(l) {
+            continue;
+        }
+        target.push_str(l);
+        target.push('\n');
+    }
+    target
+}
+
+fn bindgen() {
+    let arch = match env::var("ARCH") {
+        Ok(arch) => arch,
+        Err(_) => find_default_arch(),
+    };
+    let tuple = format!("{}-unknown-linux-gnu", arch);
+    exec(
+        cargo()
+            .env("UPDATE_BIND", "1")
+            .args(&["build", "-p", "grpcio-sys", "--target", &tuple]),
+    );
+    for f in fs::read_dir("grpc-sys/bindings").unwrap() {
+        let p = f.unwrap().path();
+        let mut content = String::new();
+        File::open(&p)
+            .unwrap()
+            .read_to_string(&mut content)
+            .unwrap();
+        let content = remove_match(&content, |l| {
+            l.starts_with("pub type ") && l.contains("= ::std::os::raw::")
+        });
+        File::create(&p)
+            .unwrap()
+            .write_all(content.as_bytes())
+            .unwrap();
+    }
+}
+
+fn cmd(c: &str) -> Command {
+    Command::new(c)
+}
+
+fn cmd_in(c: &str, dir: &str) -> Command {
+    let mut cmd = cmd(c);
+    cmd.current_dir(dir);
+    cmd
+}
+
+fn submodule() {
+    exec(cmd("git").args(&["submodule", "update", "--init", "grpc-sys/grpc"]));
+    for dir in &["cares/cares", "abseil-cpp", "re2"] {
+        exec(cmd_in("git", "grpc-sys/grpc/third_party").args(&[
+            "submodule",
+            "update",
+            "--init",
```

Comment:
Should add `--recursive`? I can't compile without the flag if I haven't done it.

Reply:
It's a for loop executed in the directory, so `recursive` is not necessary. On the other hand, recursive will also clone other unnecessary dependencies.